src/absil/il.fs

```diff
@@ -1429,114 +1562,201 @@ let convertMemberAccess (ilMemberAccess:ILMemberAccess) =
 
 let inline conditionalAdd condition flagToAdd source = if condition then source ||| flagToAdd else source &&& ~~~flagToAdd
 
-let NoMetadataIdx = -1
+type IMethodDef =
+    abstract Name: string
+    abstract Attributes: MethodAttributes
+    abstract ImplAttributes: MethodImplAttributes
+    abstract CallingConv: ILCallingConv
+    abstract Parameters: ILParameters
+    abstract Return: IReturn
+    abstract Body: ILLazyMethodBody
+    abstract SecurityDecls: ILSecurityDecls
+    abstract IsEntryPoint:bool
+    abstract GenericParams: ILGenericParameterDefs
+    abstract CustomAttrs: IAttributes 
+    abstract ParameterTypes: ILTypes
+    abstract IsIL: bool
+    abstract Code: ILCode option
+    abstract Locals: ILLocals
+    abstract IsClassInitializer: bool
+    abstract IsConstructor: bool
+    abstract IsStatic: bool
+    abstract IsNonVirtualInstance: bool
+    abstract IsVirtual: bool
+    abstract IsFinal: bool
+    abstract IsNewSlot: bool
+    abstract IsCheckAccessOnOverride: bool
+    abstract IsAbstract: bool
+    abstract MethodBody: ILMethodBody
+    abstract CallingSignature: ILCallingSignature
+    abstract Access: ILMemberAccess
+    abstract IsHideBySig: bool
+    abstract IsSpecialName: bool
+    abstract IsUnmanagedExport: bool
+    abstract IsReqSecObj: bool
+    abstract HasSecurity: bool
+    abstract IsManaged: bool
+    abstract IsForwardRef: bool
+    abstract IsInternalCall: bool
+    abstract IsPreserveSig: bool
+    abstract IsSynchronized: bool
+    abstract IsNoInline: bool
+    abstract IsAggressiveInline: bool
+    abstract IsMustRun: bool
+
+    abstract With: ?name: string * ?attributes: MethodAttributes * ?implAttributes: MethodImplAttributes * ?callingConv: ILCallingConv * 
+                 ?parameters: ILParameters * ?ret: IReturn * ?body: ILLazyMethodBody * ?securityDecls: ILSecurityDecls * ?isEntryPoint:bool * 
+                 ?genericParams: ILGenericParameterDefs * ?customAttrs: IAttributes -> IMethodDef
+
+    abstract WithSpecialName: IMethodDef
```

Comment:
I'm not that happy about these being abstract.  I think they should just be extension methods defined in terms of `With`. I'm not interested in trying to reason about implementations that give alternative implementations of these so they shouldn't be abstract